### ISSUE TITLE
Sinhala: Above base Matra positioning not proper 

### ISSUE THREAD

**Issue by Amritasingh0512** [harfbuzz/harfbuzz#967]
![sinhala_matra_positioning](https://user-images.githubusercontent.com/38288804/38623030-3c72a47c-3dc2-11e8-8d51-7a19a9a54233.PNG)

**Font used: Iskool Pota**
**Unicode value: &#x0DBB;&#x0DCA;&#x200D;&#x0DBA;&#x0DCA;&#x200D;&#x0DBA;**
**Text: ර්‍ය්‍ය**

**Comment by Amritasingh0512** [comment]
![image](https://user-images.githubusercontent.com/38288804/38623539-86ab9d86-3dc3-11e8-97e8-abf345cc0aee.png)

**Comment by ebraminio** [comment]
The difference existence is confirmed,

```
$ build/util/hb-shape iskpota.ttf -u U+0DBB,U+0DCA,U+200D,U+0DBA,U+0DCA,U+200D,U+0DBA --shaper=ot

[gid596=0+1343|gid597=0+1130]
(similar to right)

$ build/util/hb-shape iskpota.ttf -u U+0DBB,U+0DCA,U+200D,U+0DBA,U+0DCA,U+200D,U+0DBA --shaper=coretext

[gid736=3+1343|gid600=3+1130]
(similar to left)

$ WINEDLLOVERRIDES="usp10=n" wine winbuild/util/hb-shape.exe iskpota.ttf -u U+0DBB,U+0DCA,U+200D,U+0DBA,U+0DCA,U+200D,U+0DBA --shaper=uniscribe

[gid596=0+1343|gid768=4+1217|gid597=4+1130]

(hmm, I should see why)

```

**Comment by cjchapman** [comment]
The characters `\u0DBB\u0DCA\u200D` (i.e. the first three characters of the source text of this issue) are used to form the Sinhala _repaya_, which is similar to a Devanagari _reph_. Unlike the Devanagari _reph_, the Sinhala _repaya_ is no longer used in Standard Sinhala.

[_An Introduction to UNICODE for Sinhala Characters_](http://ucsc.cmb.ac.lk/ltrl/publications/uni_sin.pdf) by Samaranayake, et al. notes that:
>The “REPAYA” (ර්‍) while no longer used in standard Sinhala (and thus not included in the chart table, can be coded as the sequence “ර + ් + ZWJ” when
required.

(The newer text linked above has some missing glyph images. There is an older, scanned version in which the _repaya_ is represented more clearly [here](https://www.researchgate.net/publication/279049530_An_Introduction_to_Unicode_for_Sinhala_Characters)).

This topic of how _repaya_ should be rendered is explored further in [Noto Fonts issue #523](https://github.com/googlei18n/noto-fonts/issues/523), in which Pathum Egodawatta ([pathumego](https://github.com/pathumego)) from Sri Lanka notes that:
>Unlike Rakaransaya and Yansaya, Repaya is not in common use anymore, you wouldn't see it in literature and common use.

**Comment by behdad** [comment]
Our behavior matches Windows 7. Looks like Windows 10 changed behavior, probably as part of routing Sinhala through USE.  Investigating.

**Comment by behdad** [comment]
FWIW, this is the easy fix in the Indic shaper:
```
diff --git a/src/hb-ot-shape-complex-indic.cc b/src/hb-ot-shape-complex-indic.cc
index f08876b8..447e36cc 100644
--- a/src/hb-ot-shape-complex-indic.cc
+++ b/src/hb-ot-shape-complex-indic.cc
@@ -86,7 +86,7 @@ static const indic_config_t indic_configs[] =
   {HB_SCRIPT_KANNADA,  true, 0x0CCDu,BASE_POS_LAST, REPH_POS_AFTER_POST, REPH_MODE_IMPLICIT, BLWF_MODE_POST_ONLY},
   {HB_SCRIPT_MALAYALAM,        true, 0x0D4Du,BASE_POS_LAST, REPH_POS_AFTER_MAIN, REPH_MODE_LOG_REPHA,BLWF_MODE_PRE_AND_POST},
   {HB_SCRIPT_SINHALA,  false,0x0DCAu,BASE_POS_LAST_SINHALA,
-                                                    REPH_POS_AFTER_MAIN, REPH_MODE_EXPLICIT, BLWF_MODE_PRE_AND_POST},
+                                                    REPH_POS_AFTER_POST, REPH_MODE_EXPLICIT, BLWF_MODE_PRE_AND_POST},
 };
 
 
```

**Comment by JelleBosmaMT** [comment]
Stumbled across this issue checking if there are new issues for Noto Sinhala. Although this is closed, permit me a comment:

The sequence _Ra Virama ZWJ Ra_ is ambiguous, because _Ra Virama ZWJ_ leads to reph form and _Virama ZWJ Ra_ is for the below base form of Ra. But in this particular sequence, we want neither.
I did have a _pres_ lookup which changes _Ra Reph_ into _Ra Virama Ra_, with a multiple substitution. This solved it in the past and still solves it when testing with Safari and Edge browsers. But not in the latest Chrome. Here the Reph will be re-positioned after vowels, if present. Rephs are not supposed to be used with vowel signs, but here we should not have a reph anyway.
I managed a work around to add a context rule to the _akhn_ feature.

The easiest solution was a ligature lookup to replace _ZWJ Ra_ by _Ra_. But in Edge the Ra disappeared altogether.
Of course I can also make a ligature _Rahalant_ from _Ra Virama ZWJ_. But that gets reordered by Harfbuzz as if it is the reph. To prevent that a multiple lookup to split up the Rahalant.

Unfortunately this doesn't fix the case with split-vowels. In these cases the re-ordered vowel must be in front of the second Ra, when the _akhn_ feature is applied. I am not trying to fix that. At least the reph is in a position where it is not obscured by an ascender and it is very unlike not a real world case.
